File: py_prime_calc/prime_calc.py
#!/usr/bin/python3

import sys, time

def validate_is_prime(num, prime_list):
    # We only need to calculate up to 1/2 of the number
    # We can skip all the compount number and only use the prime as divider

    upperbound = num // 2
    i = 0
    while prime_list[i] <= upperbound:
        if num % prime_list[i] == 0:
            return False
        i += 1

    return True


def main():
    try:
        x = int(sys.argv[1])
    except (IndexError, ValueError) as error:
        x = 100000


    print("Calculate total of ", x, " elements.")
    prime_list = [2] # Using first prime '2' to initialize the list
    i = 3
    start_time = time.time()
    while i <= x:
        if validate_is_prime(i, prime_list):
            prime_list.append(i)
        i += 1
    elapsed_time = time.time() - start_time
    #print("Time=", elapsed_time, "x=", x, "Total prime=", len(prime_list), "Last prime=", prime_list[-1])
    print("  Time=", elapsed_time, "Seconds")
    print("  x=", x, "")
    print("  Total prime=", len(prime_list))
    print("  First prime=", prime_list[0])
    print("  Last prime=", prime_list[-1])

    prime_list.clear()


    print("Calculate up to ", x, " primes.")
    prime_list = [2]
    i = 3
    start_time = time.time()
    while len(prime_list) < x:
        if validate_is_prime(i, prime_list):
            prime_list.append(i)
        i += 1
    elapsed_time = time.time() - start_time
    #print("Time=", elapsed_time, "x=", x, "Total prime=", len(prime_list), "Last prime=", prime_list[-1])
    print("  Time=", elapsed_time, "Seconds")
    print("  x=", x, "")
    print("  Total prime=", len(prime_list))
    print("  First prime=", prime_list[0])
    print("  Last prime=", prime_list[-1])



if __name__ == "__main__":
    main()

File: go_prime_calc/prime_calc.go
package main

// BUG: fix the fail when command prompt argv 1 is empty
//      fix argv1 <= 0 is invalid
//      fix 

import (
    "fmt"
    "strconv"
    "os"
    "time"
    //"math"
)

func validate_is_prime(num int, prime_list[] int) bool {
    upperbound := num / 2
    i := 0
    for prime_list[i] <= upperbound {
        if num % prime_list[i] == 0 {
            return false
        }
        i += 1
    }
    return true
}


func main() {
    
    x, err := strconv.Atoi(os.Args[1])
    if err != nil { x = 100000 }

    prime_list := []int{2}
    i := 3
    start_time := time.Now()
    for i <= x {
        if validate_is_prime(i, prime_list) {
            prime_list = append(prime_list, i)
        }
        i += 1
    }
    end_time := time.Now()
    elapsed_time := end_time.Sub(start_time)
    fmt.Println("Calculated total of ", x)
    fmt.Println("  Time=", elapsed_time, "Seconds")
    fmt.Println("  x=", x)
    fmt.Println("  Total prime=", len(prime_list))
    //fmt.Println("  First prime=", prime_list[0])
    fmt.Println("  Last prime=", prime_list[len(prime_list)-1])


    prime_list = nil

    prime_list = append(prime_list, 2)
    i = 3
    start_time = time.Now()
    for len(prime_list) < x {
        if validate_is_prime(i, prime_list) {
            prime_list = append(prime_list, i)
        }
        i += 1
    }
    end_time = time.Now()
    elapsed_time = end_time.Sub(start_time)
    fmt.Println("Calculated up to ", x)
    fmt.Println("  Time=", elapsed_time, "Seconds")
    fmt.Println("  x=", x)
    fmt.Println("  Total prime=", len(prime_list))
    //fmt.Println("  First prime=", prime_list[0])
    fmt.Println("  Last prime=", prime_list[len(prime_list)-1])
}

